File: 2nd Year/2nd Semester/MPP/PROIECT-C#/Lab1/TriatlonNetworking/protobufprotocol/ProtoTriatlonProxy.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Net.Sockets;
using System.Threading;
using System.Collections.Concurrent;
using Google.Protobuf;
using log4net;
using Triatlon.Network.ProtobufProtocol;
using TriatlonNetworking.protobufprotocol;
using TriatlonServicess;

namespace Triatlon.Network.ProtobufProtocol
{
    public class ProtoTriatlonProxy : ITriatlonServices
    {
        private readonly string host;
        private readonly int port;

        private ITriatlonObserver client;
        private TcpClient connection;
        private NetworkStream stream;
        private volatile bool finished;
        private readonly BlockingCollection<Response> responses;
        private static readonly ILog logger = LogManager.GetLogger(typeof(ProtoTriatlonProxy));

        public ProtoTriatlonProxy(string host, int port)
        {
            this.host = host;
            this.port = port;
            responses = new BlockingCollection<Response>();
        }

        private void InitializeConnection()
        {
            try
            {
                connection = new TcpClient(host, port);
                stream = connection.GetStream();
                finished = false;
                StartReader();
            }
            catch (Exception e)
            {
                logger.Error($"Error connecting to server: {e.Message}");
                throw new TriatlonException("Error connecting to server: " + e.Message);
            }
        }

        private void CloseConnection()
        {
            finished = true;
            try
            {
                stream?.Close();
                connection?.Close();
                client = null;
            }
            catch (Exception e)
            {
                logger.Error($"Error closing connection: {e.Message}");
            }
        }

        private void StartReader()
        {
            Thread tw = new Thread(ReaderThread);
            tw.Start();
        }

        private void SendRequest(Request request)
        {
            try
            {
                logger.Debug($"Sending request: {request}");
                request.WriteDelimitedTo(stream);
                stream.Flush();
            }
            catch (Exception e)
            {
                logger.Error($"Error sending request: {e.Message}");
                throw new TriatlonException("Error sending request: " + e.Message);
            }
        }

        private Response ReadResponse()
        {
            try
            {
                Response response = responses.Take();
                logger.Debug($"Response received: {response}");
                if (response.Type == Response.Types.Type.Error)
                {
                    string errorMsg = response.Error;
                    logger.Error($"Error received from server: {errorMsg}");
                    throw new TriatlonException(errorMsg);
                }
                return response;
            }
            catch (Exception e)
            {
                logger.Error($"Error reading response: {e.Message}");
                throw new TriatlonException("Error reading response: " + e.Message);
            }
        }

        private bool IsUpdate(Response response)
        {
            return response.Type == Response.Types.Type.RefreeLoggedIn ||
                   response.Type == Response.Types.Type.RefreeLoggedOut ||
                   response.Type == Response.Types.Type.ResultatAdded;
        }

        private void HandleUpdate(Response response)
        {
            if (client == null) return;
            try
            {
                switch (response.Type)
                {
                    case Response.Types.Type.RefreeLoggedIn:
                        var arbitruLoggedIn = ProtoUtils.FromProto(response.Arbitru);
                        client.ArbitruLoggedIn(arbitruLoggedIn);
                        break;
                    case Response.Types.Type.RefreeLoggedOut:
                        var arbitruLoggedOut = ProtoUtils.FromProto(response.Arbitru);
                        client.ArbitruLoggedOut(arbitruLoggedOut);
                        break;
                    case Response.Types.Type.ResultatAdded:
                        var rezultatAdded = ProtoUtils.FromProto(response.Rezultat);
                        client.RezultatAdded(rezultatAdded);
                        break;
                }
            }
            catch (Exception e)
            {
                logger.Error($"Error handling update: {e.Message}");
            }
        }

        private void ReaderThread()
        {
            while (!finished)
            {
                try
                {
                    Response response = Response.Parser.ParseDelimitedFrom(stream);
                    if (response == null)
                    {
                        logger.Error("Server closed connection");
                        finished = true;
                        break;
                    }
                    logger.Debug($"Response received: {response}");
                    if (IsUpdate(response))
                    {
                        HandleUpdate(response);
                    }
                    else
                    {
                        responses.Add(response);
                    }
                }
                catch (Exception e)
                {
                    logger.Error($"Error reading response: {e.Message}");
                    finished = true;
                }
            }
        }

        // ITriatlonServices implementation

        public void Login(TriatlonModel.Arbitru arbitru, ITriatlonObserver observer)
        {
            InitializeConnection();
            client = observer;

            Request request = ProtoUtils.CreateLoginRequest(arbitru);
            SendRequest(request);
            Response response = ReadResponse();

            if (response.Type != Response.Types.Type.Ok)
            {
                CloseConnection();
                throw new TriatlonException("Login failed");
            }
        }

        public bool Register(string username, string password, string firstName, string lastName)
        {
            if (connection == null)
            {
                InitializeConnection();
            }

            TriatlonModel.Arbitru arbitru = new TriatlonModel.Arbitru(0, username, password, firstName, lastName);
            Request request = ProtoUtils.CreateRegisterRequest(arbitru);
            SendRequest(request);
            Response response = ReadResponse();

            return response.Type == Response.Types.Type.Ok;
        }

        public void Logout(TriatlonModel.Arbitru arbitru, ITriatlonObserver iTriatlonObserver)
        {
            Request request = ProtoUtils.CreateLogoutRequest(arbitru);
            SendRequest(request);
            Response response = ReadResponse();
            CloseConnection();
        }

        public void AddRezultat(TriatlonModel.Participant participant, TriatlonModel.Arbitru arbitru, TriatlonModel.TipProba tipProba, int punctaj)
        {
            Request request = ProtoUtils.CreateAddResultRequest(participant, arbitru, tipProba, punctaj);
            SendRequest(request);
            Response response = ReadResponse();

            if (response.Type != Response.Types.Type.Ok)
            {
                throw new TriatlonException("Failed to add result");
            }
        }

        public List<TriatlonModel.Rezultat> GetResultateForProba(TriatlonModel.TipProba proba)
        {
            Request request = ProtoUtils.CreateGetResultsForProbaRequest(proba);
            SendRequest(request);
            Response response = ReadResponse();

            if (response.Type == Response.Types.Type.GetResultsForProba)
            {
                List<TriatlonModel.Rezultat> rezultate = new List<TriatlonModel.Rezultat>();
                foreach (var rezultatProto in response.Rezultate)
                {
                    rezultate.Add(ProtoUtils.FromProto(rezultatProto));
                }
                return rezultate;
            }
            return new List<TriatlonModel.Rezultat>();
        }

        List<TriatlonModel.Rezultat> ITriatlonServices.GetAllResults()
        {
            try
            {
                Request request = ProtoUtils.CreateEmptyRequest(Request.Types.Type.GetAllResults);
                SendRequest(request);
                Response response = ReadResponse();

                if (response.Type == Response.Types.Type.GetAllResults)
                {
                    List<TriatlonModel.Rezultat> rezultate = new List<TriatlonModel.Rezultat>();
                    foreach (var rezultatProto in response.Rezultate)
                    {
                        rezultate.Add(ProtoUtils.FromProto(rezultatProto));
                    }
                    return rezultate;
                }
            }
            catch (Exception e)
            {
                logger.Error($"Error getting all results: {e.Message}");
            }
            return new List<TriatlonModel.Rezultat>();
        }

        List<TriatlonModel.Participant> ITriatlonServices.GetAllParticipants()
        {
            Request request = ProtoUtils.CreateEmptyRequest(Request.Types.Type.GetAllParticipants);
            SendRequest(request);
            Response response = ReadResponse();

            if (response.Type == Response.Types.Type.GetAllParticipants)
            {
                List<TriatlonModel.Participant> participants = new List<TriatlonModel.Participant>();
                foreach (var participantProto in response.Participants)
                {
                    participants.Add(ProtoUtils.FromProto(participantProto));
                }
                return participants;
            }
            return new List<TriatlonModel.Participant>();
        }

        public int CalculateTotalScore(TriatlonModel.Participant participant)
        {
            Request request = ProtoUtils.CreateCalculateTotalScoreRequest(participant);
            SendRequest(request);
            Response response = ReadResponse();

            if (response.Type == Response.Types.Type.Ok)
            {
                return response.TotalScore;
            }
            return 0;
        }

        List<TriatlonModel.Proba> ITriatlonServices.GetAllProbe()
        {
            Request request = ProtoUtils.CreateEmptyRequest(Request.Types.Type.GetAllProbe);
            SendRequest(request);
            Response response = ReadResponse();

            if (response.Type == Response.Types.Type.Ok ||
                response.Type == Response.Types.Type.GetAllProbe)
            {
                List<TriatlonModel.Proba> probe = new List<TriatlonModel.Proba>();
                foreach (var probaProto in response.Probe)
                {
                    probe.Add(ProtoUtils.FromProto(probaProto));
                }
                return probe;
            }
            return new List<TriatlonModel.Proba>();
        }

        public TriatlonModel.TipProba GetProbaForArbitru(TriatlonModel.Arbitru arbitru)
        {
            var requestBuilder = ProtoUtils.CreateEmptyRequest(Request.Types.Type.GetProbaForArbitru);

            requestBuilder.Arbitru = ProtoUtils.ToProto(arbitru);

            SendRequest(requestBuilder);
            Response response = ReadResponse();

            if (response.Type == Response.Types.Type.Ok ||
                response.Type == Response.Types.Type.GetProbaForArbitru)
            {
                if (response.TipProba != null)
                {
                    return ProtoUtils.FromProto(response.TipProba);
                }
            }
            throw new TriatlonException("No proba found for arbitru");
        }

        TriatlonModel.Arbitru ITriatlonServices.FindArbitruByUsernameAndPassword(string username, string password)
        {
            try
            {
                InitializeConnection();

                Request request = ProtoUtils.CreateFindArbitruRequest(username, password);
                SendRequest(request);
                Response response = ReadResponse();

                if (response.Type == Response.Types.Type.Ok ||
                    response.Type == Response.Types.Type.FindArbitru)
                {
                    if (response.Arbitru != null)
                    {
                        return ProtoUtils.FromProto(response.Arbitru);
                    }
                }
                return null;
            }
            catch (Exception e)
            {
                logger.Error($"Error finding arbitru: {e.Message}");
                return null;
            }
        }

    }
}

File: 2nd Year/2nd Semester/MPP/PROIECT-JAVA/Laborator/TriatlonNetworking/src/main/java/triatlon/network/protobuffprotocol/ProtoTriatlonProxy.java
package triatlon.network.protobuffprotocol;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import triatlon.model.*;
import triatlon.services.ITriatlonObserver;
import triatlon.services.ITriatlonServices;
import triatlon.services.TriatlonException;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.Socket;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.LinkedBlockingQueue;

public class ProtoTriatlonProxy implements ITriatlonServices {
    private String host;
    private int port;
    private ITriatlonObserver client;
    private Socket connection;
    private InputStream input;
    private OutputStream output;
    private BlockingQueue<TriatlonProtobufs.Response> responses;
    private volatile boolean finished;
    private static final Logger logger = LogManager.getLogger(ProtoTriatlonProxy.class);

    public ProtoTriatlonProxy(String host, int port) {
        this.host = host;
        this.port = port;
        responses = new LinkedBlockingQueue<>();
    }

    private void initializeConnection() throws TriatlonException {
        try {
            connection = new Socket(host, port);
            output = connection.getOutputStream();
            input = connection.getInputStream();
            finished = false;
            startReader();
        } catch (IOException e) {
            logger.error("Error connecting to server: " + e.getMessage());
            throw new TriatlonException("Error connecting to server: " + e.getMessage());
        }
    }

    private void closeConnection() {
        finished = true;
        try {
            if (input != null) input.close();
            if (output != null) output.close();
            if (connection != null) connection.close();
            client = null;
        } catch (IOException e) {
            logger.error("Error closing connection: " + e.getMessage());
        }
    }

    private void startReader() {
        Thread tw = new Thread(new ReaderThread());
        tw.start();
    }

    private void sendRequest(TriatlonProtobufs.Request request) throws TriatlonException {
        try {
            logger.debug("Sending request: {}", request);
            request.writeDelimitedTo(output);
            output.flush();
        } catch (IOException e) {
            logger.error("Error sending request: " + e.getMessage());
            throw new TriatlonException("Error sending request: " + e.getMessage());
        }
    }

    private TriatlonProtobufs.Response readResponse() throws TriatlonException {
        try {
            TriatlonProtobufs.Response response = responses.take();
            logger.debug("Response received: {}", response);
            if (response.getType() == TriatlonProtobufs.Response.Type.ERROR) {
                String errorMsg = response.getError();
                logger.error("Error received from server: {}", errorMsg);
                throw new TriatlonException(errorMsg);
            }
            return response;
        } catch (InterruptedException e) {
            logger.error("Error reading response: " + e.getMessage());
            throw new TriatlonException("Error reading response: " + e.getMessage());
        }
    }

    private boolean isUpdate(TriatlonProtobufs.Response response) {
        return response.getType() == TriatlonProtobufs.Response.Type.REFREE_LOGGED_IN ||
                response.getType() == TriatlonProtobufs.Response.Type.REFREE_LOGGED_OUT ||
                response.getType() == TriatlonProtobufs.Response.Type.RESULTAT_ADDED;
    }

    private void handleUpdate(TriatlonProtobufs.Response response) {
        if (client == null) return;
        try {
            switch (response.getType()) {
                case REFREE_LOGGED_IN:
                    Arbitru arbitruLoggedIn = ProtoUtils.getArbitru(response.getArbitru());
                    client.arbitruLoggedIn(arbitruLoggedIn);
                    break;
                case REFREE_LOGGED_OUT:
                    Arbitru arbitruLoggedOut = ProtoUtils.getArbitru(response.getArbitru());
                    client.arbitruLoggedOut(arbitruLoggedOut);
                    break;
                case RESULTAT_ADDED:
                    Rezultat rezultatAdded = ProtoUtils.getRezultat(response.getRezultat());
                    client.rezultatAdded(rezultatAdded);
                    break;
            }
        } catch (TriatlonException e) {
            logger.error("Error handling update: " + e.getMessage());
        }
    }

    private class ReaderThread implements Runnable {
        @Override
        public void run() {
            while (!finished) {
                try {
                    TriatlonProtobufs.Response response = TriatlonProtobufs.Response.parseDelimitedFrom(input);
                    if (response == null) {
                        logger.error("Server closed connection");
                        finished = true;
                        break;
                    }
                    logger.debug("Response received: {}", response);
                    if (isUpdate(response)) {
                        handleUpdate(response);
                    } else {
                        responses.add(response);
                    }
                } catch (IOException e) {
                    logger.error("Error reading response: " + e.getMessage());
                    finished = true;
                }
            }
        }
    }

    // ITriatlonServices implementation

    @Override
    public void login(Arbitru arbitru, ITriatlonObserver observer) throws TriatlonException {
        initializeConnection();
        this.client = observer;

        TriatlonProtobufs.Request request = ProtoUtils.createLoginRequest(arbitru);
        sendRequest(request);
        TriatlonProtobufs.Response response = readResponse();

        if (response.getType() != TriatlonProtobufs.Response.Type.OK) {
            closeConnection();
            throw new TriatlonException("Login failed");
        }
    }

    @Override
    public boolean register(String username, String password, String firstName, String lastName) throws TriatlonException {
        if (connection == null) {
            initializeConnection();
        }

        Arbitru arbitru = new Arbitru(null, username, password, firstName, lastName);
        TriatlonProtobufs.Request request = ProtoUtils.createRegisterRequest(arbitru);
        sendRequest(request);
        TriatlonProtobufs.Response response = readResponse();

        return response.getType() == TriatlonProtobufs.Response.Type.OK;
    }

    @Override
    public void logout(Arbitru arbitru, ITriatlonObserver observer) throws TriatlonException {
        TriatlonProtobufs.Request request = ProtoUtils.createLogoutRequest(arbitru);
        sendRequest(request);
        TriatlonProtobufs.Response response = readResponse();
        closeConnection();
    }

    @Override
    public void addRezultat(Participant participant, Arbitru arbitru, TipProba tipProba, int punctaj)
            throws TriatlonException {
        TriatlonProtobufs.Request request = ProtoUtils.createAddResultRequest(
                participant, arbitru, tipProba, punctaj);
        sendRequest(request);
        TriatlonProtobufs.Response response = readResponse();

        if (response.getType() != TriatlonProtobufs.Response.Type.OK) {
            throw new TriatlonException("Failed to add result");
        }
    }

    @Override
    public List<Rezultat> getResultateForProba(TipProba proba) throws TriatlonException {
        TriatlonProtobufs.Request request = ProtoUtils.createGetResultsForProbaRequest(proba);
        sendRequest(request);
        TriatlonProtobufs.Response response = readResponse();

        if (response.getType() == TriatlonProtobufs.Response.Type.GET_RESULTS_FOR_PROBA) {
            List<Rezultat> rezultate = new ArrayList<>();
            for (TriatlonProtobufs.Rezultat rezultatProto : response.getRezultateList()) {
                rezultate.add(ProtoUtils.getRezultat(rezultatProto));
            }
            return rezultate;
        }
        return new ArrayList<>();
    }

    @Override
    public List<Rezultat> getAllResults() {
        try {
            TriatlonProtobufs.Request request = ProtoUtils.createEmptyRequest(
                    TriatlonProtobufs.Request.Type.GET_ALL_RESULTS);
            sendRequest(request);
            TriatlonProtobufs.Response response = readResponse();

            if (response.getType() == TriatlonProtobufs.Response.Type.GET_ALL_RESULTS) {
                List<Rezultat> rezultate = new ArrayList<>();
                for (TriatlonProtobufs.Rezultat rezultatProto : response.getRezultateList()) {
                    rezultate.add(ProtoUtils.getRezultat(rezultatProto));
                }
                return rezultate;
            }
        } catch (TriatlonException e) {
            logger.error("Error getting all results: " + e.getMessage());
        }
        return new ArrayList<>();
    }

    @Override
    public List<Participant> getAllParticipants() throws TriatlonException {
        TriatlonProtobufs.Request request = ProtoUtils.createEmptyRequest(
                TriatlonProtobufs.Request.Type.GET_ALL_PARTICIPANTS);
        sendRequest(request);
        TriatlonProtobufs.Response response = readResponse();

        if (response.getType() == TriatlonProtobufs.Response.Type.GET_ALL_PARTICIPANTS) {
            List<Participant> participants = new ArrayList<>();
            for (TriatlonProtobufs.Participant participantProto : response.getParticipantsList()) {
                participants.add(ProtoUtils.getParticipant(participantProto));
            }
            return participants;
        }
        return new ArrayList<>();
    }

    @Override
    public int calculateTotalScore(Participant participant) throws TriatlonException {
        TriatlonProtobufs.Request request = ProtoUtils.createCalculateTotalScoreRequest(participant);
        sendRequest(request);
        TriatlonProtobufs.Response response = readResponse();

        if (response.getType() == TriatlonProtobufs.Response.Type.OK) {
            return response.getTotalScore();
        }
        return 0;
    }

    @Override
    public List<Proba> getAllProbe() throws TriatlonException {
        TriatlonProtobufs.Request request = ProtoUtils.createEmptyRequest(
                TriatlonProtobufs.Request.Type.GET_ALL_PROBE);
        sendRequest(request);
        TriatlonProtobufs.Response response = readResponse();

        if (response.getType() == TriatlonProtobufs.Response.Type.OK ||
                response.getType() == TriatlonProtobufs.Response.Type.GET_ALL_PROBE) {
            List<Proba> probe = new ArrayList<>();
            for (TriatlonProtobufs.Proba probaProto : response.getProbeList()) {
                probe.add(ProtoUtils.getProba(probaProto));
            }
            return probe;
        }
        return new ArrayList<>();
    }

    @Override
    public TipProba getProbaForArbitru(Arbitru arbitru) throws TriatlonException {
        TriatlonProtobufs.Request request = ProtoUtils.createEmptyRequest(
                TriatlonProtobufs.Request.Type.GET_PROBA_FOR_ARBITRU);
        request = request.toBuilder()
                .setArbitru(ProtoUtils.createArbitru(arbitru))
                .build();
        sendRequest(request);
        TriatlonProtobufs.Response response = readResponse();

        if (response.getType() == TriatlonProtobufs.Response.Type.OK ||
                response.getType() == TriatlonProtobufs.Response.Type.GET_PROBA_FOR_ARBITRU) {
            if (response.hasTipProba()) {
                return ProtoUtils.getTipProba(response.getTipProba());
            }
        }
        return null;
    }

    @Override
    public Arbitru findArbitruByUsernameAndPassword(String username, String password) {
        try {
            initializeConnection();

            TriatlonProtobufs.Request request = ProtoUtils.createFindArbitruRequest(username, password);
            sendRequest(request);
            TriatlonProtobufs.Response response = readResponse();

            if (response.getType() == TriatlonProtobufs.Response.Type.OK ||
                    response.getType() == TriatlonProtobufs.Response.Type.FIND_ARBITRU) {
                if (response.hasArbitru()) {
                    return ProtoUtils.getArbitru(response.getArbitru());
                }
            }
            return null;
        } catch (TriatlonException e) {
            logger.error("Error finding arbitru: {}", e.getMessage());
            return null;
        }
    }
}